astropy/table/operations.py

```diff
@@ -767,9 +769,19 @@ def _join(left, right, keys=None, join_type='inner',
     if isinstance(_col_name_map, Mapping):
         _col_name_map.update(col_name_map)
 
+    unmask_table_columns(out)
+
     return out
 
 
+def unmask_table_columns(tbl):
+    if tbl.masked:
+        tbl._set_masked(False)
+        for col in tbl.itercols():
+            if isinstance(col, MaskedColumn) and not np.any(col.mask):
+                tbl[col.name] = Column(col, copy=False)
```

Comment:
It looks like `tbl[col.name] = tbl[col.name].view(Column)` works the same and is 3 times faster (91 vs 248 us).  Should that be fine @mhvk?

Reply:
Indeed, what we really want here is the unmasked data; in `MaskedArray`, this is exposed as both `data` and `_data`, and all that does is `self.view(self._baseclass)`. Now, we cannot use this directly since the `_baseclass` is `BaseColumn`, not `Column` (not completely sure why...). But that means your view does the right thing.

The one thing that this does not deal with is possible subclasses. Maybe easiest is to just let those fend for themselves (i.e., do `type(col) is MaskedColumn`, or `type(col._baseclass) is BaseColumn).

Though, a more general question: is there a risk that this change of nature of the column will surprise users? They have a table with some masked columns, do something, and suddenly the `.mask` property on their column disappears?

Reply:
@mhvk - those are good points that need consideration.  Esp. the last one, for instance hstack'ing two same-length tables and having your masked column turn unmasked.  That does seem less than perfect.  OTOH right now the opposite can happen in a join or vstack where unmasked columns become masked for no good reason.  But there at least you don't lose functionality.

This probably helps you understand why I don't like the whole two-class solution to masking..  There are pros and cons, but the cons are not trivial.  If / when the `MaskedQuantity` class comes back to life as a duck array we will have a fair bit of pain in the table code supporting it.  Quantity won't be a generic mixin, but will need special handling.  We may need to somehow generalize the Column/MaskedColumn dichotomy to allow for multiple Class / MaskedClass variants.  Or something...

Reply:
(In case it wasn't clear, I'm thinking about better solutions in the mean time.)

Reply:
I've been thinking on and off about the `Masked` versions and very much see your point that it is a pain to have different classes. I've still got some hope for a generic `info.mask` attribute - as long as we keep this super-simple (e.g., reductions would be lead to masked values if even one element was masked). But that's for another discussion!

Reply:
OK, new implementation of the `join` update that doesn't use such a big hammer.  I'm much happier with this.  Now I have a path for the other ones, but each operation type will be different.